sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PushPartialAggregationThroughJoin.scala

```diff
@@ -0,0 +1,102 @@
+
+/**
+ * Push down the partial aggregation through join if it cannot be planned as broadcast hash join.
+ */
+object PushPartialAggregationThroughJoin extends Rule[LogicalPlan]
+  with PredicateHelper
+  with JoinSelectionHelper {
+
+  private def isSupportPushdown(aggregateExpressions: Seq[NamedExpression]): Boolean = {
+    aggregateExpressions.forall {
+      case _: Attribute => true
+      case Alias(_: Attribute, _) => true
+      case Alias(_: Literal, _) => true
+      case _ => false
+    }
+  }
+
+  private def split(expressions: Seq[NamedExpression], left: LogicalPlan, right: LogicalPlan) = {
+    val (leftExpressions, rest) =
+      expressions.partition(e => e.references.nonEmpty && e.references.subsetOf(left.outputSet))
+    val (rightExpressions, remainingExpressions) =
+      rest.partition(e => e.references.nonEmpty && e.references.subsetOf(right.outputSet))
+
+    (leftExpressions, rightExpressions, remainingExpressions)
+  }
+
+  private def pushdown(
+      agg: Aggregate,
+      leftKeys: Seq[Expression],
+      rightKeys: Seq[Expression],
+      join: Join): LogicalPlan = {
+    val (leftGroupExps, rightGroupExps, remainingGroupExps) =
+      split(agg.groupingExpressions.map(_.asInstanceOf[NamedExpression]), join.left, join.right)
+
+    val (leftAggExps, rightAggExps, remainingAggExps) =
+      split(agg.aggregateExpressions, join.left, join.right)
+
+    val newLeftGroup = leftGroupExps ++ leftKeys.flatMap(_.references)
+    val newLeftAgg: Seq[NamedExpression] = newLeftGroup ++ leftAggExps.flatMap(_.references)
+
+    val newRightGroup = rightGroupExps ++ rightKeys.flatMap(_.references)
+    val newRightAgg: Seq[NamedExpression] = newRightGroup ++ rightAggExps.flatMap(_.references)
+
+    val newJoin = join.copy(
+      left = Aggregate(newLeftGroup.distinct, newLeftAgg.distinct, true, join.left),
+      right = Aggregate(newRightGroup.distinct, newRightAgg.distinct, true, join.right))
+
+    if (agg.isPartialOnly) {
+      Project(agg.aggregateExpressions, newJoin)
+    } else {
+      agg.copy(child = newJoin)
+    }
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan.transformWithPruning(
+    _.containsAllPatterns(AGGREGATE, JOIN), ruleId) {
+    case agg @ Aggregate(_, _, _, join: Join)
+        if join.left.isInstanceOf[Aggregate] || join.right.isInstanceOf[Aggregate] =>
+      agg
+    case agg @ Aggregate(_, _, _, Project(_, join: Join))
+        if join.left.isInstanceOf[Aggregate] || join.right.isInstanceOf[Aggregate] =>
+      agg
+
+    case agg @ Aggregate(_, aggExprs, _,
+          join @ ExtractEquiJoinKeys(Inner, leftKeys, rightKeys, None, _, left, right, _))
+        if isSupportPushdown(aggExprs) && (leftKeys ++ rightKeys).nonEmpty &&
+          !canPlanAsBroadcastHashJoin(join, conf) =>
+      pushdown(agg, leftKeys, rightKeys, join)
+
+    case agg @ Aggregate(_, aggExprs, _, Project(projectList,
+          join @ ExtractEquiJoinKeys(Inner, leftKeys, rightKeys, None, _, left, right, _)))
```

Comment:
[question] can we extend this to LeftSemi / LeftAnti as well, as we can very well have a parent agg on non-join key, then in this case we could push partialAgg of aggExpr, left join-key down in the left side of join to reduce our join input. 
As per my understanding it might  be possible, Am I am missing something here ?

Reply:
We can support it step by step.